ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java

```diff
@@ -0,0 +1,276 @@
+      this.listener = listener;
+    }
+
+    /**
+     * Tell the background executor to perform any initialization needed. This should be called after registering this
+     * listener using {@link HelixPropertyStore#subscribe}. This will ensure that the first event sent to
+     * {@link #listener} will contain the entire set of current configs.
+     */
+    void start() {
+      eventExecutor.execute(this::initializeIfNeeded);
+    }
+
+    @Override
+    public void onDataChange(String path) {
+      LOGGER.debug("DataNodeConfig path {} changed", path);
+      onPathChange(path);
+    }
+
+    @Override
+    public void onDataCreate(String path) {
+      LOGGER.debug("DataNodeConfig path {} created", path);
+      onPathChange(path);
+    }
+
+    @Override
+    public void onDataDelete(String path) {
+      // TODO handle node deletions dynamically. Doing so requires further work in the ClusterChangeHandler impl
+      LOGGER.info("DataNodeConfig path {} deleted. This requires a restart to handle", path);
+    }
+
+    private void onPathChange(String path) {
+      eventExecutor.execute(() -> updateOrInitialize(path));
+    }
+
+    private Iterable<DataNodeConfig> lazyIterable(Collection<ZNRecord> records) {
+      return () -> records.stream().map(converter::convert).filter(Objects::nonNull).iterator();
+    }
+
+    /**
+     * Either perform a full initialization or notify the listener about an incremental update to a path. This should be
+     * called from the event executor thread.
+     * @param changedPath the path that changed.
+     */
+    private void updateOrInitialize(String changedPath) {
+      lock.lock();
+      try {
+        if (!initializeIfNeeded()) {
+          ZNRecord record = propertyStore.get(changedPath, null, AccessOption.PERSISTENT);
+          if (record != null) {
+            listener.onDataNodeConfigChange(lazyIterable(Collections.singleton(record)));
+          } else {
+            LOGGER.info("DataNodeConfig at path {} not found", changedPath);
+          }
+        }
+      } catch (Exception e) {
+        LOGGER.error("Exception during DataNodeConfig change", e);
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    /**
+     * If initialization has not yet been completed succesfully, attempt an initialization. This should be called from
+     * the event executor thread.
+     * @return {@code true} if a successful full initialization (whether successful or not) was attempted in this call.
+     */
+    private boolean initializeIfNeeded() {
+      lock.lock();
+      try {
+        if (initialized) {
+          return false;
+        }
+        List<ZNRecord> records = propertyStore.getChildren(CONFIG_PATH, null, AccessOption.PERSISTENT);
+        listener.onDataNodeConfigChange(lazyIterable(records));
+        initialized = true;
+      } catch (Exception e) {
+        LOGGER.error("Exception during DataNodeConfig initialization", e);
```

Comment:
Just curious, if exception occurs, how can we block the startup of current node?

Reply:
Hmm, that is a good point. So, with InstanceConfig listeners, does the exception thrown in the addListener call bubble up to DatacenterInitializer? If so, I will have to address this. Also, I realize that this method does not currently block until initialization is finished. I may have to make changes to support this.

Reply:
Changed the add listener call to wait for initialization to finish in the background thread to match how the other DataNodeConfigSource works. The code will now rethrow an exception thrown by the listener during initialization.